notif.go

```diff
@@ -29,61 +23,51 @@ func (nn *netNotifiee) Connected(n inet.Network, v inet.Conn) {
 	default:
 	}
 
-	dht.plk.Lock()
-	defer dht.plk.Unlock()
-
-	conn, ok := nn.peers[v.RemotePeer()]
-	if ok {
-		conn.refcount++
+	p := v.RemotePeer()
+	protos, err := dht.peerstore.SupportsProtocols(p, dhtProtocols...)
+	if err == nil && len(protos) != 0 {
+		dht.plk.Lock()
```

Comment:
what exactly does this lock protect here?

Reply:
Seems to protect Update -- but that's a public interface function. 
It has no business being both public and requiring the lock to be held.

Reply:
This particular lock is probably unnecessary because `Connect` and `Disconnect` notifications are synchronous (although I still want to leave it as it doesn't hurt and I like being consistent). However, we *do* need to take it below in the `Disconnect` handler and in `testConnection`. Otherwise, we could end up with the following interleaving:

| testConnection | Disconnect |
|---------------------------------------------|------|
| Observe that we are connected  | |
|                                                     | Observe that we are disconnected (disconnect event happens) |
|                                                    | Remove peer from routing table |
| Add peer to routing table            |        |

This is an alternative to reference counting open connections (what we did before) that doesn't require keeping a bunch of additional state.

Reply:
ok, fair enough. Can we add a comment to that effect -- it looks totally out of place otherwise.

Reply:
Good point. Done.

Reply:
What would be the effect of calling `Update` without this lock? I am concerned about the public interface uses of it.